lib/puppetx/simp/compliance_mapper.rb

```diff
@@ -374,135 +374,181 @@ def import(filename, data)
-                          current["telemetry"] = current["telemetry"] + specification["telemetry"]
-                        else
-                          retval[parameter] = specification["settings"].merge(specification)
+
+                  confine.each do |confinement_setting, confinement_value|
+                    if confinement_setting == 'module_name'
+                      unless @callback.module_list.map { |obj| obj['name'] }.include?(confinement_value)
+                        delete_item = true
+                        throw :confine_end
+                      end
+
+                      if confine['module_version']
+                        require 'semantic_puppet'
+                        known_module = @callback.module_list.select { |obj| obj['name'] == confinement_setting }
+
+                        currentver = SemanticPuppet::Version.parse(known_module.first['version'])
+                        requiredver = SemanticPuppet::VersionRange.parse(confine['module_version'])
+
+                        unless requiredver.include?(currentver)
+                          delete_item = true
+                          throw :confine_end
                         end
                       end
                     end
+
+                    fact_value = @callback.lookup_fact(confinement_setting)
+                    unless fact_value == confinement_value
+                      delete_item = true
+                      throw :confine_end
+                    end
                   end
                 end
               end
             end
+
+            delete_item
           end
-          retval
-        end
-      end
-    end
+
+          # If we didn't find anything, we can just bail
+          return retval if specifications.empty?
+
+          if specifications.count > 1
+            @callback.debug("WARN: Multiple valid specifications found for #{specifications.first['settings']['parameter']}, they will be merged in the order that they were defined")
+          end
+
+          specifications.each do |specification|
+            parameter = specification['settings']['parameter']
+
+            # Process all entries that have passed the confinement checks and
+            # return the appropriate match
+            if retval.key?(parameter)
+              # Merge
+              # XXX ToDo: Need merge settings support
+              current = retval[parameter]
+
+              specification['settings'].each do |key, value|
+                unless key == 'parameter'
+                  case current[key].class.to_s
+                  when 'Array'
+                    current[key] = (current[key] + Array(value)).uniq
+                  when 'Hash'
+                    current[key].merge!(value)
+                  else
+                    current[key] = Marshal.load(Marshal.dump(value))
+                  end
+                end
+              end
+
+              current['telemetry'] = current['telemetry'] + specification['telemetry']
```

Comment:
Should retval be set instead of current?  Otherwise, I don't see the reason for doing the current logic.

Reply:
It's because the value is getting manipulated above. Honestly, this was copied from the previous implementation, so you may be right but I didn't want to change more than was absolutely necessary.

Reply:
This needs to be completely rewritten, we can handle this at that time.